### PR TITLE
Propagation rule for ternary operation

### PR DESCRIPTION
**Pull request by jjsjann123**
What's in this PR:
* Abstract the propagation rule for 2 operands to N operands with `resolveAllocationOrder`;
    * Adding propagation rule for ternary op using the refactored `resolveAllocationOrder`;
* Added a propagation rule for pad to just forward input allocation order to output;
* A cpp test for ternary ops;

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 1]
```diff
@@ -72,6 +74,22 @@ class AllocationOrderInferencer : public IterVisitor {
...
+  //   3. has the highest number of non_broadcast IterDomain
+  //
+  // The function is used to resolve allocation order propagation for operator
+  // with multiple operands. The one operand with the most number of
+  // non-broadcast IterDomain will be dominating the output allocation order.
+  // The motivation behind it to avoid breaking allocation order propagation
+  // from operands produced by broadcast. e.g. When a binary operator could take
+  // in a channels_last 4d tensor and an unsqueezed bias vector. We'll want to
+  // propagate the channels_last allocation order to output.
+  TensorView* resolveAllocationOrder(const std::vector<Val*>& candidates);
```
Comment (jjsjann123) [inline csrc/preseg_passes/allocation_order_inference.cpp:91]: This is merely a refactor to remove the special short-cut for binary operation. Idea is basically the same that we want to propagate the *biggest* tensor's allocation order to output.

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 2]
```diff
@@ -166,69 +221,26 @@ void AllocationOrderInferencer::handle(BroadcastOp* op) {
...
-  // otherwise, we propagate the one with more non-broadcast iterdomains.
-  if (countNonBroadcastID(lhs) >= countNonBroadcastID(rhs)) {
-    alloc_order_map_[out] = lhs_iter->second;
-  } else {
-    alloc_order_map_[out] = rhs_iter->second;
-  }
+void AllocationOrderInferencer::handle(PadOp* op) {
+  auto* out = dynamic_cast<TensorView*>(op->out());
+  auto* in = dynamic_cast<TensorView*>(op->in());
+  propagateAllocationOrder(in, out);
```
Comment (jjsjann123) [inline csrc/preseg_passes/allocation_order_inference.cpp:243]: pad is trivial and *almost* identical to unary. so I didn't bother adding a test.

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 3]
```diff
@@ -72,6 +74,22 @@ class AllocationOrderInferencer : public IterVisitor {
     return false;
   }
 
+  // returns the candidate that dominates the allocation order.
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:77]:
```suggestion
  // Returns the candidate operand that dominates the allocation order.
```

File: csrc/preseg_passes/allocation_order_inference.cpp (hunk 1 above)
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:91]:
```suggestion
  //
  // Pre-condition: `candidates` must be the input operands of the same Expr. 
  TensorView* resolveAllocationOrder(const std::vector<Val*>& candidates);
```

Am I right?

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 4]
```diff
@@ -72,6 +74,22 @@ class AllocationOrderInferencer : public IterVisitor {
...
+  //
+  // It scans through each candidate to find the first one that:
+  //   1. is a TensorView
+  //   2. has an entry in alloc_order_map_
+  //   3. has the highest number of non_broadcast IterDomain
+  //
+  // The function is used to resolve allocation order propagation for operator
+  // with multiple operands. The one operand with the most number of
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:85]:
```suggestion
  // with multiple operands. The operand with the most number of
```

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 5]
```diff
@@ -82,6 +100,43 @@ class AllocationOrderInferencer : public IterVisitor {
...
+  // helper utils to count the number of non broadcast iterdomain
+  auto countNonBroadcastID = [](const TensorView* tv) -> size_t {
+    return std::count_if(
+        tv->getMaybeRFactorDomain().begin(),
+        tv->getMaybeRFactorDomain().end(),
+        [&](auto ptr_id) { return !ptr_id->isBroadcast(); });
+  };
+
+  for (auto* val_ptr : candidates) {
+    auto* tv_ptr = dynamic_cast<TensorView*>(val_ptr);
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:117]:
```suggestion
  for (auto* val : candidates) {
    auto* tv = dynamic_cast<TensorView*>(val);
```

`auto*` already indicates they are pointers, so we don't need `_ptr` in the name.

**Review by wujingyue — APPROVED** [review]

**Comment by wujingyue** [comment]
Thanks -- nice refactor! 🚢

**Pull request merged**